minidump-processor/src/op_analysis.rs

```diff
@@ -154,222 +234,341 @@ mod amd64 {
+                    | Opcode::RDTSC
+                    | Opcode::RDTSCP
+                    | Opcode::RETF
+                    | Opcode::STI
+                    | Opcode::SWAPGS
+                    | Opcode::SYSEXIT
+                    | Opcode::SYSRET
+                    | Opcode::VMCALL
+                    | Opcode::VMCLEAR
+                    | Opcode::VMLAUNCH
+                    | Opcode::VMPTRLD
+                    | Opcode::VMPTRST
+                    | Opcode::VMREAD
+                    | Opcode::VMRESUME
+                    | Opcode::VMWRITE
+                    | Opcode::VMXOFF
+                    | Opcode::VMXON
+                    | Opcode::WBINVD
+                    | Opcode::WRMSR
+                    | Opcode::XSETBV
+            )
         }
+    }
 
+    impl MemoryAccess {
         /// Determine the memory accesses implied by the given instruction and context
         ///
-        /// Uses the instruction provided in `decoded_instruction` and the given registers in
-        /// `context` to determine information about memory accessed by the given instruction.
-        ///
         /// # Errors
         ///
         /// The most likely cause of an error is that a register named by the given instruction
-        /// is invalid, so determining the memory access for the instruction is not possible.
-        pub fn get_instruction_memory_access(
-            self,
-            decoded_instruction: Instruction,
-        ) -> Result<Vec<MemoryAccess>, OpAnalysisError> {
-            let mut memory_accesses = Vec::new();
-            self.direct(&mut memory_accesses, &decoded_instruction)?;
-            self.indirect(&mut memory_accesses, &decoded_instruction)?;
-            Ok(memory_accesses)
+        /// is invalid.
+        fn from_instruction(
+            instruction: Instruction,
+            context: &MinidumpContext,
+        ) -> Result<Vec<Self>, OpAnalysisError> {
+            let mut accesses = Vec::new();
+            if let Some(opcode) = CommonOpcode::from_amd64_opcode(instruction.opcode()) {
+                Self::add_common_opcode_accesses(&mut accesses, opcode, instruction, context)?;
+            } else {
+                Self::add_uncommon_opcode_accesses(&mut accesses, instruction, context)?;
+            }
+            Ok(accesses)
         }
 
-        fn direct(
-            &self,
+        fn add_common_opcode_accesses(
             accesses: &mut Vec<MemoryAccess>,
-            decoded_instruction: &Instruction,
+            opcode: CommonOpcode,
+            instruction: Instruction,
+            context: &MinidumpContext,
         ) -> Result<(), OpAnalysisError> {
-            // Shortcut -- If the instruction doesn't access memory, just return an empty list
-            let mem_size = match decoded_instruction.mem_size() {
+            // Shortcut -- If the instruction doesn't access memory, just return
+            let mem_size = match instruction.mem_size() {
```

Comment:
Hmm... I have a concern here -- I'm not totally sure that Yaxpeax will account for implicit memory accesses with this method. If not, then you will exit too early here without considering implicit accesses.

Perhaps you could check that this does what you want it to?

Reply:
Yaxpeax gives `mem_size` for `ret`, `call`, `push` and `pop` and the likes, so this should be fine. For some reason it gives `mem_size` for `retf` but not `iretd`. I'll further investigate.

Reply:
So it seems that for yaxpeaax doesn't give `mem_size` for `iretd` and `iretq`, but does so for the other common opcodes that have implicit access 🤔. I'm also not sure if it even makes sense here to claim that `iret` access the memory pointed by the instruction pointer.